doc/04-Upgrading.md

```diff
@@ -3,6 +3,64 @@
+time to perform the reindexing on the `history` table even if it is not blocked. Nonetheless, you can use the
+`show processlist` command to determine whether an excessive number of queries have been stuck in a waiting state.
+
+```
+MariaDB [icingadb]> show processlist;
++------+-----+-----+----------+-----+------+---------------------------------+------------------------------------+-----+
+| Id   | ... | ... | db       | ... | Time | State                           | Info                               | ... |
++------+-----+-----+----------+-----+------+---------------------------------+------------------------------------+-----+
+| 1475 | ... | ... | icingadb | ... | 1222 | Waiting for table metadata lock | INSERT INTO "notification_history" | ... |
+| 1485 | ... | ... | icingadb | ... | 1262 | Creating sort index             | SELECT history.id, history....     | ... |
+| 1494 | ... | ... | icingadb | ... | 1224 | Waiting for table metadata lock | ALTER TABLE history ADD INDEX ...  | ... |
+| 1499 | ... | ... | icingadb | ... | 1215 | Waiting for table metadata lock | INSERT INTO "notification_history" | ... |
+| 1500 | ... | ... | icingadb | ... | 1215 | Waiting for table metadata lock | INSERT INTO "state_history" ...    | ... |
++------+-----+-----+----------+-----+------+---------------------------------+------------------------------------+-----+
+39 rows in set (0.001 sec)
+```
+
+In the above output you can see that there are approximately `39` queries (`39 rows in set (0.001 sec)`), including the
+`ALTER TABLE history ADD INDEX` query from the upgrade script, waiting for a metadata lock, they're just minimised to
+the bare essentials. Unfortunately, only one of these queries is holding the `table metadata lock` that everyone else
+is now waiting for, which in this case is a `SELECT` statement initiated by Icinga DB Web in the `icingadb/history`
+view, which takes an unimaginably long time. Note that there might be multiple `SELECT` statements started before the
+upgrade script in your case when the history view of your Icinga DB Web is opened by different Icinga Web users at the
+same time.
+
+You can now either just wait for the `SELECT` statements to finish by themselves and let them block the upgrade script
+and all Icinga DB queries on all `*_history` tables or forcibly terminate them and let the remaining queries do their
+work. In this case, cancelling that one blocking `SELECT` query will let the upgrade script continue normally without
+blocking any other queries.
+``` 
+MariaDB [icingadb]> kill 1485;
+```
+
+In case you are insecure about which Icinga DB Web queries are blocking, you may simply cancel all long-running
+`SELECT` statements listed with `show processlist` (see column `Time`). Cancelling a `SELECT` query is usually not that
+disastrous, so feel free to abort every single one of them matching the Icinga DB database (see column `db`).
+
```

Comment:
This was all very detailed, but limited to MySQL-like databases. As this also affects PostgreSQL, this should also be covered.

Reply:
Fair point. Have you tried how exactly this behaves with PostgreSQL? At least with PostgreSQL≥13 and the history view, I don't expect such slow queries (see https://github.com/Icinga/icingadb/pull/686#issuecomment-1988043508). But I don't know what exactly will happen with older versions (or other slow queries if you run them manually).

Reply:
As PostgreSQL 12 is still being [supported till end of this year](https://www.postgresql.org/support/versioning/), this might be something to at least mention. Otherwise, when there's no issue with a version greater 12, I would say to just add one paragraph mentioning potential issues with PostgreSQL 12 and advising people to upgrade, and mentioning that newer versions should be fine. Otherwise PostgreSQL users might be unsure to upgrade, as there is such a long text regarding potential issues with MySQL.

Reply:
**Before:**
```pgsql
icingadb=# \c timing on
Timing is on.
icingadb=# SELECT COUNT(*) FROM history;
 count  
--------
 405399
(1 row)

Time: 33.537 ms

icingadb=# SELECT 1 as name FROM history ORDER BY event_time, event_type LIMIT 1;
 name 
------
    1
(1 row)

Time: 38.162 ms
```

**After:**
```pgsql
icingadb=# SELECT COUNT(*) FROM history;
 count  
--------
 414913
(1 row)

Time: 20.501 ms
icingadb=# SELECT 1 as name FROM history ORDER BY event_time, event_type LIMIT 1;
 name 
------
    1
(1 row)

Time: 0.382 ms
```

Reply:
```bash
root@713a54a05873:/# psql --version
psql (PostgreSQL) 12.18 (Debian 12.18-1.pgdg120+2)
```

Reply:
That shows that the extended index indeed provides a benefit for older PostgreSQL versions. However, I think this thread is about whether a situation similar to "Waiting for table metadata lock" can happen with PostgreSQL. But so far, I don't know if anyone actually tried to verify this. It could also be an option that it handles the metadata sufficiently different than MySQL so that it doesn't have a similar locking problem.

Reply:
If we create the index that https://github.com/Icinga/icingadb/pull/686#discussion_r1549825905 way, shouldn't this discussion become irrelevant then? If PostgreSQL says that it won't hold any locks, there shouldn't be such a situation as with MySQL.

Reply:
> If PostgreSQL says that it won't hold any locks

Does it say so? https://www.postgresql.org/docs/current/sql-createindex.html says for `CONCURRENTLY`:

> When this option is used, PostgreSQL will build the index without taking any locks that prevent concurrent inserts, updates, or deletes on the table; whereas a standard index build locks out writes (but not reads) on the table until it's done.

That doesn't say anything about metadata locking, where this glitch happens with MySQL. PostgreSQL isn't trying to reimplement MySQL, so that doesn't mean it must have the same problem. So far, the only reason I see for this actually being an issue with PostgreSQL is the claim "As this also affects PostgreSQL" at the beginning of the discussion but my question in https://github.com/Icinga/icingadb/pull/686#discussion_r1549709572 whether this was actually observed, wasn't answered so far (@oxzi).

Reply:
Let me rephrase my original comment: The update also affects PostgreSQL, not just MySQL and its derivatives. So if I, as a potential PostgreSQL user, would scroll through the update logs and see lots of details about potential problems with MySQL, but not a single word about PostgreSQL, I might be afraid to apply the schema update.

As the warning above this paragraph now (or was it already before and I just missed it?) starts with a note regarding PostgreSQL, my objections are now void.